hedera-node/hedera-app/src/main/java/com/hedera/node/app/signature/impl/SignatureVerificationFutureImpl.java

```diff
@@ -0,0 +1,192 @@
+        if (future != null) {
+            return future.cancel(mayInterruptIfRunning);
+        }
+
+        return true;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public boolean isCancelled() {
+        return canceled;
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * Due to the asynchronous nature of {@link TransactionSignature}, it is possible that at the time this method
+     * is called, one or more {@link TransactionSignature}s do not yet have a {@link Future}. If that is the case,
+     * then we are clearly not done!
+     */
+    @Override
+    public boolean isDone() {
+        if (canceled) {
+            return true;
+        }
+
+        final var future = txSig.getFuture();
+        return future != null && future.isDone();
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * Blocks on each {@link Future}. Due to the asynchronous nature of {@link TransactionSignature}, it is possible
+     * that a {@link Future} has not been assigned to one or more {@link TransactionSignature}s. In that case, we will
+     * wait for the {@link Future} before proceeding.
+     */
+    @NonNull
+    @Override
+    public SignatureVerification get() throws InterruptedException, ExecutionException {
+        txSig.waitForFuture().get(); // Wait for the future to be assigned and completed
+        return new SignatureVerificationImpl(key, hollowAccount, txSig.getSignatureStatus() == VALID);
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * Blocks on each {@link Future}. Due to the asynchronous nature of {@link TransactionSignature}, it is possible
+     * that a {@link Future} has not been assigned to one or more {@link TransactionSignature}s. In that case, we will
+     * wait for the {@link Future} before proceeding (up to the timeout).
+     */
+    @NonNull
+    @Override
+    public SignatureVerification get(long timeout, TimeUnit unit)
+            throws InterruptedException, ExecutionException, TimeoutException {
+        var time = System.currentTimeMillis();
+        var millisRemaining = unit.toMillis(timeout);
+        while (txSig.getFuture() == null) {
+            final var now = System.currentTimeMillis();
+            millisRemaining -= (now - time);
+            if (millisRemaining <= 0) {
+                // If there was no time left, then TimeoutException
+                throw new TimeoutException("Timed out waiting for signature verification to complete");
+            } else {
+                // If there was time left, but we didn't yet have a Future on the TransactionSignature, then
+                // rather than blocking in a tight loop (like we do for "get"), we have to just wait a bit and
+                // check for timeout and try again. That way, we can still time out.
+                //noinspection BusyWait
+                Thread.sleep(1);
+            }
+        }
+        txSig.getFuture().get(millisRemaining, TimeUnit.MILLISECONDS); // Wait for the future to complete
```

Comment:
You are right. I rewrote it per @jsync-swirlds suggestion:

```
    public SignatureVerification get(long timeout, TimeUnit unit)
            throws InterruptedException, ExecutionException, TimeoutException {
        final long deadline = System.currentTimeMillis() + unit.toMillis(timeout);
        while (txSig.getFuture() == null) {
            if (System.currentTimeMillis() > deadline) {
                // If there was no time left, then TimeoutException
                throw new TimeoutException("Timed out waiting for signature verification to complete");
            } else {
                // If there was time left, but we didn't yet have a Future on the TransactionSignature, then
                // rather than blocking in a tight loop (like we do for "get"), we have to just wait a bit and
                // check for timeout and try again. That way, we can still time out.
                //noinspection BusyWait
                Thread.sleep(1);
            }
        }

        // Wait for the future to complete
        txSig.getFuture().get(deadline - System.currentTimeMillis(), TimeUnit.MILLISECONDS);
        return new SignatureVerificationImpl(key, hollowAccount, txSig.getSignatureStatus() == VALID);
    }
```

This seems simpler and I think addresses the concern.